lib/credentials/assume_role_with_web_identity_credentials.js

```diff
@@ -0,0 +1,153 @@
+      return {
+        envTokenFile: process.env[ENV_TOKEN_FILE],
+        roleArn: process.env[ENV_ROLE_ARN],
+        roleSessionName: process.env['AWS_IAM_ROLE_SESSION_NAME']
+      };
+    }
+  },
+
+  /**
+   * Returns params from shared config variables
+   *
+   * @api private
+   */
+  getParamsFromSharedConfig: function getParamsFromSharedConfig() {
+    try {
+      var profiles = {};
+      var profilesFromConfig = {};
+      if (process.env[AWS.util.configOptInEnv]) {
+        profilesFromConfig = AWS.util.iniLoader.loadFrom({
+          isConfig: true,
+          filename: process.env[AWS.util.sharedConfigFileEnv]
+        });
+      }
+      var profilesFromCreds = AWS.util.iniLoader.loadFrom({
+        filename: (process.env[AWS.util.configOptInEnv] && process.env[AWS.util.sharedCredentialsFileEnv])
+      });
+      for (var i = 0, profileNames = Object.keys(profilesFromConfig); i < profileNames.length; i++) {
+        profiles[profileNames[i]] = profilesFromConfig[profileNames[i]];
+      }
+      for (var i = 0, profileNames = Object.keys(profilesFromCreds); i < profileNames.length; i++) {
+        profiles[profileNames[i]] = profilesFromCreds[profileNames[i]];
+      }
+      var profile = profiles[process.env.AWS_PROFILE || AWS.util.defaultProfile] || {};
+
+      if (Object.keys(profile).length === 0) {
+        throw AWS.util.error(
+          new Error('Profile ' + this.profile + ' not found'),
+          { code: 'AssumeRoleWithWebIdentityCredentialsProviderFailure' }
+        );
+      }
+
+      return {
+        envTokenFile: profile['web_identity_token_file'],
+        roleArn: profile['role_arn'],
+        roleSessionName: profile['role_session_name']
+      };
+    } catch (err) {
+      throw err;
+    }
+  },
+
+  /**
+   * Refreshes credentials using {AWS.STS.assumeRoleWithWebIdentity}
+   *
+   * @callback callback function(err)
+   *   Called when the STS service responds (or fails). When
+   *   this callback is called with no error, it means that the credentials
+   *   information has been loaded into the object (as the `accessKeyId`,
+   *   `secretAccessKey`, and `sessionToken` properties).
+   *   @param err [Error] if an error occurred, this value will be filled
+   * @see AWS.Credentials.get
+   */
+  refresh: function refresh(callback) {
+    this.coalesceRefresh(callback || AWS.util.fn.callback);
+  },
+
+  /**
+   * @api private
+   */
+  getRandomSessionName: function getRandomSessionName() {
+    return crypto.randomBytes(64).toString('base64');
+  },
+
+  /**
+   * @api private
+   */
+  load: function load(callback) {
+    var self = this;
+    try {
+      var params = self.getParamsFromEnv();
+      if (!params) {
+        params = self.getParamsFromSharedConfig();
+      }
+      if (params) {
+        var oidcToken = fs.readFileSync(params.envTokenFile, {encoding: 'ascii'});
+        var webIdentityCredentials = new AWS.WebIdentityCredentials({
```

Comment:
Do we need to create a new `WebIdentityCredentials` instance at every load? Now it may create an STS client [down the hood](https://github.com/aws/aws-sdk-js/blob/2627d1af01bf1a1dc745361c3edf9e825594dcd4/lib/credentials/web_identity_credentials.js#L93). Can we keep a reference of a single `WebIdentityCredentials` instance?

Reply:
`WebIdentityCredentials` doesn't have an option to update `WebIdentityToken` other than passing it in the constructor. In [STS documentation](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/STS.html) too, I didn't find an option to update `WebIdentityToken` after the STS object is created.

We can save reference to `WebIdentityToken`, and create new WebIdentityCredentials object only if the token is updated. What do you think?

Reply:
Because `WebIdentityCredentials` sets the `WebIdentityToken` parameters through the servcie configuration's `params` entry, see [here](https://github.com/aws/aws-sdk-js/blob/master/lib/credentials/web_identity_credentials.js#L110). For more information, see [here](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/creating-and-calling-service-objects.html#specifying-service-object-parameters). 

In fact, you can access the parameter by:

```javascript
webIdentityCredentials.service.config.params.WebIdentityToken = oidcToken
```

Reply:
Resolved in https://github.com/aws/aws-sdk-js/pull/2737/commits/bf420078b194427a7389ff72c992ac6f0b93d67d